functional-programming/src/main.ts

```diff
@@ -0,0 +1,84 @@
+import { Either, fromPromise, ap, right, getOrElse, flatten, left } from "./fp/either";
+import { pipe } from "./fp/utils";
+import { fetchClient, fetchExecutor } from "./fetching";
+import { ClientUser, Demand, ExecutorUser } from "./types";
+import { fold, fromNullable } from "./fp/maybe";
+import { distance } from "./utils";
+import { fromCompare, ordNumber, revert } from "./fp/ord";
+
+type Response<R> = Promise<Either<string, R>>
+
+const getExecutor = (): Response<ExecutorUser> => fromPromise(fetchExecutor());
+const getClients = (): Response<Array<ClientUser>> => fromPromise(fetchClient()
+  .then(value => value.map(({ demands, ...restParams }) => ({
+    demands: fromNullable(demands),
+    ...restParams
+  }))));
+
+export enum SortBy {
+  distance = "distance",
+  reward = "reward",
+}
+
+export const show = (sortBy: SortBy) => (clients: Array<ClientUser>) => (executor: ExecutorUser): Either<string, string> => {
+  const rewardPredicate = (item1: ClientUser, item2: ClientUser) => (
+    ordNumber.compare(item1.reward, item2.reward)
+  );
+
+  const distancePredicate = (item1: ClientUser, item2: ClientUser) => (
+    ordNumber.compare(distance(item1.position, executor.position), distance(item2.position, executor.position))
+  );
+
+  const rewardOrd = revert(fromCompare(rewardPredicate));
+  const distanceOrd = fromCompare(distancePredicate);
+
+  const compareClients = (a: ClientUser, b: ClientUser): number => {
+    return sortBy === SortBy.reward ? rewardOrd.compare(a, b) : distanceOrd.compare(a, b);
+  };
+
+  const sortedClients = [...clients].sort(compareClients);
+
+  const executorClients = sortedClients.filter(client => {
+    const isMeetDemands = fold<Array<Demand>, boolean>(
+      () => true,
+      (clientDemands) => clientDemands.every(demand => executor.possibilities.includes(demand))
+    );
+    return isMeetDemands(client.demands);
+  });
+
+  const clientsCount = executorClients.length;
+  const allClientsCount = sortedClients.length;
+
+  let result = `This executor meets the demands of only ${clientsCount} out of ${allClientsCount} clients\n\n`;
```

Comment:
it's mutable variable to build the final output string.
In FP, it's generally preferable to avoid mutable variables and instead use immutable data structures and functions that return new data.
In this case, the result variable could be replaced with a function or functions that constructs the output string and returns it.